### PR TITLE
spike: dashboard templates

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

I made a dashboard template in #12343. It was manual and difficult

And now @joethreepwood has asked for two more templates in #12732

This would be slow, fiddly and so error-prone.

Quoting Kent Beck: "for each desired change, make the change easy (warning: this may be hard), then make the easy change" (https://twitter.com/kentbeck/status/250733358307500032?lang=en)

## Changes

* adds dashboard templates as a concept
* lets you save a dashboard templates from a dashboard (behind a flag)
* adds a tab to the dashboards page to show templates and let you rename and delete them (behind a flag)

Once this is merged the two templates in #12732 can be saved as templates and then copied into a migration

dashboard-templates

## How did you test this code?

* added developer tests
* ran it locally
* cypress tests

### CODE REVIEW

File: frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        },
+        currentTab: [
+            DashboardsTab.All as DashboardsTab,
+            {
+                setCurrentTab: (_, { tab }) => tab,
+            },
+        ],
+    },
+    selectors: {
+        filteredDashboardTemplates: [
```
Comment (pauldambra) [inline frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts:40]: GitHub really needs semantic diff. This file shows as all added but it's only this selector that has been added

File: frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        ],
+        dashboards: [
+            (selectors) => [
+                dashboardsModel.selectors.nameSortedDashboards,
+                selectors.searchTerm,
+                selectors.currentTab,
+                selectors.user,
+            ],
+            (dashboards, searchTerm, currentTab, user): DashboardType[] => {
+                let listToFilter: DashboardType[]
```
Comment (pauldambra) [inline frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts:62]: and this variable has been renamed

File: frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+                if (currentTab === DashboardsTab.Templates) {
```
Comment (pauldambra) [inline frontend/src/scenes/dashboard/dashboards/dashboardsLogic.ts:64]: and this guard for the new tab

File: frontend/src/scenes/dashboard/newDashboardLogic.ts [hunk 4]
```diff
@@ -77,6 +78,9 @@ export const newDashboardLogic = kea<newDashboardLogicType>([
             actions.submitNewDashboard()
         },
         showNewDashboardModal: () => {
+            actions.getAllDashboardTemplates()
```
Comment (pauldambra) [inline frontend/src/scenes/dashboard/newDashboardLogic.ts:81]: don't reset the modal when opening it, or you can't set values on the form when opening it

File: posthog/api/test/dashboards/test_dashboard.py [hunk 5]
```diff
@@ -548,14 +548,6 @@ def test_dashboard_item_layout(self):
 
         self.assertTrue("lg" in first_tile_layouts)
 
-    def test_dashboard_from_template(self):
```
Comment (pauldambra) [inline posthog/api/test/dashboards/test_dashboard.py:551]: deletions in this file are because the tests are moved to a new file

**Comment by andyvan-ph** [comment]
Sounds good. Feature suggestion... Templates for insights as well. 😁

**Comment by pauldambra** [comment]
exporting-templates

**Review by mariusandra — APPROVED** [review]
Sweet!

That's a lot of code though 😅. I definitely didn't follow every line of every test in detail. Overall looks good to me, just some small things:

1. The "save as template" menu option is there if you don't have the flag enabled.
2. In this interface:
image

- the first column is really wide... not sure what to do here if anything
- when we have just one export option, let's just call it "export as json" without the second level?

3. The first button "Create dashboard using this template" needs more line height to breathe. This is way too cramped. A two-line text button can't be the same height as a online text button

**Comment by pauldambra** [comment]
@mariusandra thinking more on the "import from github"

Shall I do the laziest thing and dump the entire JSON into one field in the DB? 

Occurs to me I can then make a unique key on that field...... 🤔

**Comment by mariusandra** [comment]
> @mariusandra thinking more on the "import from github"
> 
> Shall I do the laziest thing and dump the entire JSON into one field in the DB?
> 
> Occurs to me I can then make a unique key on that field...... 🤔

You can't get lazier than this:

https://github.com/PostHog/posthog/blob/d6bc434f98f67db06acf4fcf71845862e50c4783/posthog/api/plugin.py#L289-L293

... and it has worked just fine so far, so totally up to you. Don't over think it, it's a MVP after all. Later on we can make an optimised `https://respository.posthog.com/dashboards.json` endpoint, but for now, 🤷

**Comment by pauldambra** [comment]
* users can save templates
    * in project or organization scope
    * dependent on permissions
* users can export templates 
* and import templates
* when a project is created a dashboard is created using a template
    * so at least one template has to be present before a team is created
    * so one template is included in the original migration that creates the templates

* we need to get templates into self-hosted instances
* and we want some templates to be global for cloud

other global templates are stored in a git repo

celery periodically retrieves templates from that repo and stores them
users can trigger a synchronous refresh from the repo

**Comment by pauldambra** [comment]
## global templates can't be renamed or deleted

Screenshot 2022-11-24 at 18 28 18

## organization templates can't be renamed or deleted by non-org admins

Screenshot 2022-11-24 at 18 29 12

## project templates can be renamed or deleted

Screenshot 2022-11-24 at 18 28 38

**Comment by pauldambra** [comment]
global templates are here: https://github.com/PostHog/templates-repository

File: frontend/src/scenes/dashboard/dashboardTemplates/dashboardTemplateLogic.tsx [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+type TilePayload = InsightTilePayload | TextTilePayload
+
+export interface DashboardTemplateRequest {
+    dashboard_name: string
+    dashboard_description: string
+    tags: string[]
+    tiles: TilePayload[]
+}
+
+export async function pollTemplateRefreshStatus(task_id: string): Promise<void> {
```
Comment (mariusandra) [inline frontend/src/scenes/dashboard/dashboardTemplates/dashboardTemplateLogic.tsx:45]: This feels like a lot of complexity. Imagine time goes on and we have 30 dashboard templates. I understand in that case why we need to do this in a task, but... why would we want to automatically download and install all of them? I imagined the "repository" to work similar to plugins -> you get a list of templates, and click "download" next to the templates you care about. You know, like a repository :).

And in this case, if the only thing the backend does is to refresh one JSON file (and store it in redis like we do for plugins), all this queying and fetching falls off.

**Review by mariusandra — CHANGES_REQUESTED** [review]
So, as we spoke, it'd be great to make this into a system similar to plugins, where we just download one `.json`, show it in an interface with a bunch of "install template" buttons. 

Rough and raw comment below and inline.

---------

I'm not really sure what should happen here:

image

Restarting Celery made this happen:

image

... and it's not clear what are my dashboards, what got added, etc. I also have no idea what do the new templates contain -> the label "website traffic" gives me limited info. 

Something that would help is a description line under the template name. Alternatively, "preview mode" that shows what the template contains before you create a real tempalte with it.

I think this doesn't need to be a submenu:

image

Exports fail for me:


image

I did `pip install -r requirements.txt` and restarted celery, but that didn't bring those itertools into the game.

**Pull request closed without merging**